Pr fir implmentation

zacck opened the pull request:
Adds a Filter to the Audio Signals

zacck committed:
Add config options for filters

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
Define number of taps our filter uses

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
fir_filter: compute filtering of audio samples

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

StaryAnoda commented on src/audio_service.c:
```diff
@@ -11,6 +10,66 @@ K_MEM_SLAB_DEFINE_STATIC(mem_slab, BLOCK_SIZE, BLOCK_COUNT, 4);
+float32_t samples_f[SAMPLES_PER_BLOCK];
+float32_t filtered_samples[SAMPLES_PER_BLOCK];
+
+const float32_t filter_taps[53] = {
```
```suggestion
const float32_t filter_taps[NUM_TAPS] = {
```

zacck replied:
thank  you!

StaryAnoda commented on src/audio_service.c:
```diff
@@ -11,6 +10,66 @@ K_MEM_SLAB_DEFINE_STATIC(mem_slab, BLOCK_SIZE, BLOCK_COUNT, 4);
 
+static float32_t fir_state[SAMPLES_PER_BLOCK + NUM_TAPS - 1 ];
+float32_t samples_f[SAMPLES_PER_BLOCK];
+float32_t filtered_samples[SAMPLES_PER_BLOCK];
```
```suggestion
static float32_t filtered_samples[SAMPLES_PER_BLOCK];
```

zacck replied:
Hmmm, I am not sure I want this as static, because if it is copied it can be in 2 places and space is expensive

StaryAnoda commented on src/audio_service.c:
```diff
@@ -11,6 +10,66 @@ K_MEM_SLAB_DEFINE_STATIC(mem_slab, BLOCK_SIZE, BLOCK_COUNT, 4);
 static const struct device *i2s_dev;
 
+static float32_t fir_state[SAMPLES_PER_BLOCK + NUM_TAPS - 1 ];
+float32_t samples_f[SAMPLES_PER_BLOCK];
```
```suggestion
static float32_t samples_f[SAMPLES_PER_BLOCK];
```

zacck replied:
these as well

StaryAnoda commented on src/audio_service.c:
```diff
@@ -11,6 +10,66 @@ K_MEM_SLAB_DEFINE_STATIC(mem_slab, BLOCK_SIZE, BLOCK_COUNT, 4);
 
 static const struct device *i2s_dev;
 
+static float32_t fir_state[SAMPLES_PER_BLOCK + NUM_TAPS - 1 ];
```
```suggestion
static float32_t fir_state[SAMPLES_PER_BLOCK + NUM_TAPS - 1];
```

StaryAnoda commented on src/audio_service.c:
```diff
@@ -58,14 +124,33 @@ void audio_sense_thread(void *arg1, void *arg2, void *arg3)
 		if (ret == 0) {
-			for (int b = 0; b < 3; b++) {
-				q31_t rms_q31;
+			for (int b = 1; b <= 3; b++) {
```
Maybe I don't see the whole picture, but why do we need this loop?

StaryAnoda replied:
In this implementation we call `k_mem_slab_free(&mem_slab, (void *)mem_block);` three times. Double free problem.

zacck replied:
Oh this one is hard won, remove it and see :) 

But really this is how the i2s driver works and each loop is a single block each block is 100ms. 
One can’t have just one block in a way this is close to the mininum

zacck committed:
inference: add inference thread

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
Start moving fitering to inference thread

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
inference: initialize memblock and copy to it well

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
Add reviews of code

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
fsm_service: Add an infering state and add supporting events and transitions

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
Audio infering changes

- added a mem slab to use for inference
- added a  fifo to use for sending block addresses
- we started posting events to the state machine

zacck committed:
Start building how we will do inference

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
inference: clear correct mem_block and signal inference end

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
inference: handle inference end signal

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
Move filtering to inferencing and run it there

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
fft_signal: Activate FFT CMSIS lib

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
fft_signal: move to 4096 samples per a block for fft

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
fft_signal: broken fft computation of signal

Signed-off-by: Zacck Osiemo <coderv63@gmail.com>

zacck committed:
Inference: Fixes

- check status of cmsis fft instance
- compute fft of signal and hold it in the right buffer
